common/task/interface.go

```diff
@@ -18,22 +18,42 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
+//go:generate mockgen -copyright_file ../../LICENSE -package $GOPACKAGE -source $GOFILE -destination interface_mock.go -self_package github.com/uber/cadence/common/task
+
 package task
 
 import (
 	"github.com/uber/cadence/common"
 )
 
 type (
-	// SequentialTaskProcessor is the generic coroutine pool interface
-	// which process sequential task
-	SequentialTaskProcessor interface {
+	// Processor is the generic coroutine pool interface
+	// which process tasks
+	Processor interface {
+		common.Daemon
+		Submit(task Task) error
+	}
+
+	// PriorityTaskProcessor is the generic coroutine pool interface
+	// which process tasks with priority
+	PriorityTaskProcessor interface {
+		common.Daemon
+		Submit(task Task, priority int) error
+	}
+
+	// PriorityScheduler is the generic interface for scheduling
+	// tasks with priority
+	PriorityScheduler interface {
```

Comment:
- Can we call this TaskScheduler ? Scheduler name is too broad
 - From an interface design perspective, I would suggest to keep the Scheduler interface simple and move the priority to Task interface. Also Consume don't belong inside the Scheduler interface (see comments later). Something along the lines of :

```
TaskScheduler interface {
   Submit(Task task) error
}

Task interface {
    ...
   Priority() int
}
```
This way, priority scheduler simply becomes a type of Scheduler implementation which makes use of task_priority. There could be other scheduler implementations which ignore priority (like today).